### PR TITLE
[SPARK-48049][BUILD] Upgrade Scala to 2.13.14

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to upgrade `scala` from `2.13.13` to `2.13.14`.

### Why are the changes needed?
https://github.com/scala/scala/releases/tag/v2.13.14
image

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by panbingkun** [comment]
Let's wait for it
https://contributors.scala-lang.org/t/scala-2-13-14-release-planning/6581/12
image

**Review by dongjoon-hyun — COMMENTED** [review]
Wow. Nice timing. Thank you, @panbingkun !

**Comment by dongjoon-hyun** [comment]
Like we observed before, I guess we need to wait new compatible plugins and `Ammonite` shell additionally. Isn't it?

cc @LuciferYang , too

**Comment by panbingkun** [comment]
> Like we observed before, I guess we need to wait new compatible plugins and `Ammonite` shell additionally. Isn't it?
> 
> cc @LuciferYang , too

Yeah, that's right. 😄
I guess it should be able to catch up with spark4.0 release.

**Review by dongjoon-hyun — COMMENTED** [review]
For the record, this is blocked by
- https://github.com/lightbend/genjavadoc/issues/355

**Comment by dongjoon-hyun** [comment]
Thank you so much, @SethTisue !

**Comment by dongjoon-hyun** [comment]
Although we are waiting for `Ammonite` still, could you base this PR once more, @panbingkun ?
- https://github.com/com-lihaoyi/Ammonite/pull/1482

**Comment by panbingkun** [comment]
> Although we are waiting for `Ammonite` still, could you base this PR once more, @panbingkun ?
> 
> * [Add support for Scala 2.13.14 com-lihaoyi/Ammonite#1482](https://github.com/com-lihaoyi/Ammonite/pull/1482)

Sure, done.

**Comment by panbingkun** [comment]
> https://repo1.maven.org/maven2/com/typesafe/genjavadoc/genjavadoc-plugin_2.13.14/0.19/

I have updated the version of `genjavadoc` in the file `project/SparkBuild.scala`.

**Comment by dongjoon-hyun** [comment]
Thanks!

**Review by dongjoon-hyun — COMMENTED** [review]
Just a question. CI seems to complain the following. Is this related to this PR?
```
Error:  Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.4.1:enforce (enforce-versions) on project spark-tools_2.13: 
Error:  Rule 3: org.codehaus.mojo.extraenforcer.dependencies.EnforceBytecodeVersion failed with message:
Error:  Found Banned Dependency: org.jline:jline:jar:3.25.1
```

**Comment by panbingkun** [comment]
> jline

I guess it may be related to the fact that the `Ammonite` supporting `scala 2.13.14` has not been released.

File: connector/connect/client/jvm/pom.xml [hunk 1]
```diff
@@ -73,7 +73,7 @@
     </dependency>
     <dependency>
       <groupId>com.lihaoyi</groupId>
-      <artifactId>ammonite_${scala.version}</artifactId>
+      <artifactId>ammonite_2.13.13</artifactId>
```
Comment (panbingkun) [inline connector/connect/client/jvm/pom.xml:76]: Only for test
After the `Ammonite` supports `scala 2.13.14`, I will restore it.

Comment (dongjoon-hyun) [reply connector/connect/client/jvm/pom.xml:76]: Yes, it's merged ten minutes ago. Looking forward to seeing the new release.

**Comment by dongjoon-hyun** [comment]
Although we already use `ammonite_2.13.13` without this jline issue, let's wait and see~
> I guess it may be related to the fact that the Ammonite supporting scala 2.13.14 has not been released.

**Comment by panbingkun** [comment]
@SethTisue 
Just right, you are here too
https://github.com/scala/scala/pull/10717
https://github.com/scala/scala/pull/10717/files#diff-87e6fe6a086cec9825e72357099f0ae6b84800ea025c5aee97b6cc7088c6cd4d

I found that in `scala-2.13.14`, the version of `org.jline:jline` was upgraded from `3.14.1` to `3.15.1` because:
https://github.com/scala/bug/issues/12957
image

But I found that `org.jline:jline:3.15.1` is compiled based on `JDK 21`. 
image
So the maven plugin  `maven-enforcer-plugin` helped us find this issue when Spark upgraded `the scala version` from `2.13.13` to `2.13.14`.
Because our current Spark master branch supports `JDK 17` and `JDK 21`.
Based on the above reasons, I would like to ask if it is also `reasonable` and `feasible` for us to exclude `org.jline:jline:3.15.1` from `org.scala-lang:scala-compiler:jar:2.13.14`,  and explicitly let it use `org.jline:jline:3.14.1`?
image

**Comment by LuciferYang** [comment]
https://github.com/jline/jline3/blob/c75301facc8716b59c1d57d3e3c5943358022560/pom.xml#L93-L95

https://github.com/jline/jline3/blob/c75301facc8716b59c1d57d3e3c5943358022560/pom.xml#L422-L434

The release version of `jline 3.25.1` is still Java 8, this should not affect Spark's use of it.

**Comment by LuciferYang** [comment]
[INFO] Restricted to JDK 17 yet org.jline:jline:jar:3.25.1:compile contains org/jline/terminal/impl/ffm/CLibrary$termios.class targeted to 65.-257


Oh... It seems that `CLibrary$termios.class` is not compatible with Java 8.

**Comment by dongjoon-hyun** [comment]
Thank you for the analysis and sharing, @panbingkun and @LuciferYang .

**Comment by panbingkun** [comment]
> Thank you for the analysis and sharing, @panbingkun and @LuciferYang .



> [INFO] Restricted to JDK 17 yet org.jline:jline:jar:3.25.1:compile contains org/jline/terminal/impl/ffm/CLibrary$termios.class targeted to 65.-257
> 
> Oh... It seems that `CLibrary$termios.class` is not compatible with Java 8.
> 
> In jline 3.25.1, CLibrary references some implementations that were moved to the `java.lang.foreign` package in Java 21, but in Java 17, they were still in the `jdk.incubator.foreign` package.
> 
> However, in jline 3.24.1, CLibrary also import classes in `java.lang.foreign` package, it's just that the final jline jar does not include the `terminal-ffm` module, so it won't trigger a compilation error from the `maven-enforcer-plugin`.
> 
> So it seems that using jline 3.25.1 should not pose any issues. It's unclear whether it's possible to workaround this by modifying the rules of the `maven-enforcer-plugin`. @panbingkun

Let me try, We should be able to refer to the following:
https://github.com/mojohaus/extra-enforcer-rules/issues/46
image

But I'm not sure we really want to do this? 
We have 2 options:
1. Use `org.jline:jline:3.14.1` (exclude `org.jline:jline:3.15.1` from `org.scala-lang:scala-compiler:jar:2.13.14`)
2. Use `org.jline:jline:3.15.1` (Modify the rule `enforceBytecodeVersion` by configuring `ignoreClasses ` and bypass it ) 
WDYT @dongjoon-hyun

**Comment by panbingkun** [comment]
Currently, this GA has turned green.

**Comment by dongjoon-hyun** [comment]
Do we have a corresponding issue in the Scala community or JLine community (or GitHub issue)?

IMO,
a. We can ignore Scala 2.13.14 completely due to this issue.
b. We can cut the transitive dependency as you mentioned as Option 1. However, it's `Scala`'s dependency. So, the downstream projects or users (who are not aware of this) will complain again and again.
c. We cannot choose Option 2 because the downstream projects or users will be affected and complain again.

For Apache Spark `4.0.0-preview`, let's choose (a) because we didn't have enough time to play around with this new Scala version.

Until `4.0.0` release, we had better keep tracking the upstream community (Scala/JLine) activity about this topic. The best case is for them to fix it for all downstream (including us).

**Comment by LuciferYang** [comment]
> > Thank you for the analysis and sharing, @panbingkun and @LuciferYang .
> 
> > [INFO] Restricted to JDK 17 yet org.jline:jline:jar:3.25.1:compile contains org/jline/terminal/impl/ffm/CLibrary$termios.class targeted to 65.-257
> > Oh... It seems that `CLibrary$termios.class` is not compatible with Java 8.
> > In jline 3.25.1, CLibrary references some implementations that were moved to the `java.lang.foreign` package in Java 21, but in Java 17, they were still in the `jdk.incubator.foreign` package.
> > However, in jline 3.24.1, CLibrary also import classes in `java.lang.foreign` package, it's just that the final jline jar does not include the `terminal-ffm` module, so it won't trigger a compilation error from the `maven-enforcer-plugin`.
> > So it seems that using jline 3.25.1 should not pose any issues. It's unclear whether it's possible to workaround this by modifying the rules of the `maven-enforcer-plugin`. @panbingkun
> 
> Let me try, We should be able to refer to the following: [mojohaus/extra-enforcer-rules#46](https://github.com/mojohaus/extra-enforcer-rules/issues/46) image
> 
> But I'm not sure we really want to do this? We have 2 options:
> 
> 1. Use `org.jline:jline:3.14.1` (exclude `org.jline:jline:3.15.1` from `org.scala-lang:scala-compiler:jar:2.13.14`)
> 2. Use `org.jline:jline:3.15.1` (Modify the rule `enforceBytecodeVersion` by configuring `ignoreClasses ` and bypass it)
> 
> WDYT @dongjoon-hyun

If it can be confirmed that Scala 2.13.14 does not use this class(`CLibrary$termios`), I think it can be done this way because , but is this the only issue?

**Comment by LuciferYang** [comment]
> Do we have a corresponding issue in the Scala community or JLine community (or GitHub issue)?
> 
> IMO, a. We can ignore Scala 2.13.14 completely due to this issue. b. We can cut the transitive dependency as you mentioned as Option 1. However, it's `Scala`'s dependency. So, the downstream projects or users (who are not aware of this) will complain again and again. c. We cannot choose Option 2 because the downstream projects or users will be affected and complain again.
> 
> For Apache Spark `4.0.0-preview`, let's choose (a) because we didn't have enough time to play around with this new Scala version.
> 
> Until `4.0.0` release, we had better keep tracking the upstream community (Scala/JLine) activity about this topic. The best case is for them to fix it for all downstream (including us).

+1 for choice  Option a if there are no critical issues that need to be fixed by Scala 2.13.14

one issue I can see is [Fix ArrayBuilder regression in Scala 2.13.13 (OutOfMemoryError when adding empty arrays) ](https://github.com/scala/scala/pull/10722)

Let's create an issue in the Scala community? @panbingkun

**Comment by panbingkun** [comment]
> Do we have a corresponding issue in the Scala community or JLine community (or GitHub issue)?
> 
> IMO, a. We can ignore Scala 2.13.14 completely due to this issue. b. We can cut the transitive dependency as you mentioned as Option 1. However, it's `Scala`'s dependency. So, the downstream projects or users (who are not aware of this) will complain again and again. c. We cannot choose Option 2 because the downstream projects or users will be affected and complain again.
> 
> For Apache Spark `4.0.0-preview`, let's choose (a) because we didn't have enough time to play around with this new Scala version.
> 
> Until `4.0.0` release, we had better keep tracking the upstream community (Scala/JLine) activity about this topic. The best case is for them to fix it for all downstream (including us).

@dongjoon-hyun @LuciferYang 
Let's keep an eye on it. I have submitted this issue to the `scala community`
https://github.com/scala/bug/issues/12994
Thank all!

**Comment by panbingkun** [comment]
@dongjoon-hyun @LuciferYang 
Based on the detailed explanation of https://github.com/scala/bug/issues/12994, 
image

and `double-check` by `debug jline` on local env. The verification process is as follows:
1.Based on this pr upgrade (the version of `scala` is `2.13.14` , the version of `jline` is `3.25.1`)
compile spark by commands:
```
./build/mvn -Pyarn -Pkubernetes -Pvolcano -Phive -Phive-thriftserver -Phadoop-cloud -Pspark-ganglia-lgpl -Pkinesis-asl clean package -DskipTests
```
image

2.When open `spark-shell` with `the debug mode` on local , and we will see the following run `thread stack`
```
export SPARK_SUBMIT_OPTS=-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=8888
```

3.The code logic of `org.jline:jline:3.25.1` is as follows:
https://github.com/jline/jline3/blob/jline-parent-3.25.1/terminal/src/main/java/org/jline/terminal/TerminalBuilder.java#L654-L675
image

When loading `org.jline.terminal.ffm`, it will automatically `skip` because of the exception 
```
java.lang.UnsupportedClassVersionError: org/jline/terminal/impl/ffm/FfmTerminalProvider has been compiled by a more recent version of the Java Runtime (class file version 65.65535), this version of the Java Runtime only recognizes class file versions up to 61.0
``` 
so we can `completely ignore` it in the maven plugin `maven-enforcer-plugin#enforceBytecodeVersion`, because it will automatically `skip` because of `loading failure` in the lower version of JDK (JDK < 21).

File: pom.xml [hunk 2]
```diff
@@ -2981,6 +2981,9 @@
                     <maxJdkVersion>${java.version}</maxJdkVersion>
                     <ignoredScopes>test</ignoredScopes>
                     <ignoredScopes>provided</ignoredScopes>
+                    <ignoreClasses>
+                      <ignoreClass>org.jline.terminal.impl.ffm.*</ignoreClass>
```
Comment (dongjoon-hyun) [inline pom.xml:2985]: Please add a comment with the pointer https://github.com/scala/bug/issues/12994 .

**Comment by dongjoon-hyun** [comment]
Thank you, @panbingkun .

**Comment by dongjoon-hyun** [comment]
- Is the PR ready?
- `ammonite` 3.0.0-M1` is enough?
    - https://ammonite.io/#3.0.0-M1

**Comment by LuciferYang** [comment]
I think we  need wait a new Ammonite release version, as 3.0-M1 does not support Scala 2.13.14.

image

**Comment by LuciferYang** [comment]
> > * Is the PR ready?
> > * `ammonite` 3.0.0-M1` is enough?
> >   
> >   * https://ammonite.io/#3.0.0-M1
> 
> I use this version `[3.0.0-M1-19-a7973e17](https://mvnrepository.com/artifact/com.lihaoyi/ammonite_3.3.3/3.0.0-M1-19-a7973e17)`, which currently supports `scala 2.13.14` https://mvnrepository.com/artifact/com.lihaoyi/ammonite image

This seems to be a snapshot or test version, not an official release？

**Comment by panbingkun** [comment]
> > > * Is the PR ready?
> > > * `ammonite` 3.0.0-M1` is enough?
> > >   
> > >   * https://ammonite.io/#3.0.0-M1
> > 
> > 
> > I use this version `[3.0.0-M1-19-a7973e17](https://mvnrepository.com/artifact/com.lihaoyi/ammonite_3.3.3/3.0.0-M1-19-a7973e17)`, which currently supports `scala 2.13.14` https://mvnrepository.com/artifact/com.lihaoyi/ammonite image
> 
> This seems to be a snapshot or test version, not an official release？

https://github.com/com-lihaoyi/Ammonite/issues/1481#issuecomment-2097731736

> > > * Is the PR ready?
> > > * `ammonite` 3.0.0-M1` is enough?
> > >   
> > >   * https://ammonite.io/#3.0.0-M1
> > 
> > 
> > I use this version `[3.0.0-M1-19-a7973e17](https://mvnrepository.com/artifact/com.lihaoyi/ammonite_3.3.3/3.0.0-M1-19-a7973e17)`, which currently supports `scala 2.13.14` https://mvnrepository.com/artifact/com.lihaoyi/ammonite image
> 
> This seems to be a snapshot or test version, not an official release？


I have asked the community to release an official version for us at his convenience
https://github.com/com-lihaoyi/Ammonite/issues/1481#issuecomment-2097731736

**Comment by LuciferYang** [comment]
https://github.com/com-lihaoyi/Ammonite/releases/tag/3.0.0-M2 

3.0.0-M2 released ~ @panbingkun

**Comment by panbingkun** [comment]
> https://github.com/com-lihaoyi/Ammonite/releases/tag/3.0.0-M2
> 
> 3.0.0-M2 released ~ @panbingkun

Thanks~ ❤️
Updated.

File: project/SparkBuild.scala [hunk 3]
```diff
@@ -266,7 +266,7 @@ object SparkBuild extends PomBuild {
       .orElse(sys.props.get("java.home").map { p => new File(p).getParentFile().getAbsolutePath() })
       .map(file),
     publishMavenStyle := true,
-    unidocGenjavadocVersion := "0.18",
+    unidocGenjavadocVersion := "0.19",
```
Comment (LuciferYang) [inline project/SparkBuild.scala:269]: Please upgrade `GenJavadoc ` in a separate pr first. @panbingkun

Comment (panbingkun) [reply project/SparkBuild.scala]: Done

Comment (panbingkun) [reply project/SparkBuild.scala]: @LuciferYang 
I think we can upgrade scala to `2.13.14` only after https://github.com/apache/spark/pull/46579 is merged.
1.
https://mvnrepository.com/artifact/com.typesafe.genjavadoc/genjavadoc-plugin
image
image

2.
https://github.com/panbingkun/spark/actions/runs/9079936342/job/24951167973
image

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @panbingkun and @LuciferYang and all.
Merged to master for Apache Spark 4.0.0-preview2-rc2.

**Pull request closed without merging**